tests/spiped_nc/graceful_shutdown.c

```diff
@@ -0,0 +1,96 @@
+#include <signal.h>
+#include <stdlib.h>
+
+#include "events.h"
+#include "warnp.h"
+
+#include "graceful_shutdown.h"
+
+struct graceful_shutdown_cookie {
+	void (* begin_shutdown)(void *);
+	void * caller_cookie;
+};
+
+static volatile sig_atomic_t should_shutdown = 0;
+static void * graceful_shutdown_timer_cookie = NULL;
```

Comment:
Why have this here, but the cookie and callback as part of a structure which gets allocated on the heap, anyway?